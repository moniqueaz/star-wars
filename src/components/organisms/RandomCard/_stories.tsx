import { Story, Meta } from '@storybook/react/types-6-0';
import RandomCard from '.';

export default {
  title: 'Organisms/RandomCard',
  component: RandomCard,
} as Meta;

export const Default: Story = () => {
  return <RandomCard count={3} slug="people" />;
};
