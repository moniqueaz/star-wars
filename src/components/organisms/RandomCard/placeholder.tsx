import styled from 'styled-components';
import Skeleton from 'components/atoms/Skeleton';

const Placeholder = () => {
  return <Wrapper>
    <Skeleton />
  </Wrapper>;
};

export default Placeholder;

const Wrapper = styled.div`
  width: 25rem;
  height: 46rem;
`;
