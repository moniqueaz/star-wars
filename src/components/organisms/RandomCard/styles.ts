import styled, { css } from 'styled-components';

export const Wrapper = styled.div``;

export const Content = styled.div`
  ${( { theme } ) => css`
    margin-bottom: ${theme.spacings.small};
    
    @media(min-width: '1024px'){
      margin-bottom: ${theme.spacings.large};
    }
  `}
`;

