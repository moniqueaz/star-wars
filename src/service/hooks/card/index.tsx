import useSWR from 'swr';
import { fetcher } from 'utils/helpers';
import config from 'infra/config';

export const useCard = (slug: string | string[], id: number) => {
  const { data, error, isValidating } = useSWR(
      `${config.getOne}/${slug}/${id}`,
      fetcher,
  );

  const excludedList = ['name', 'homeworld', 'films', 'species', 'vehicles', 'starships', 'created', 'edited', 'url', 'residents', 'characters', 'planets', 'opening_crawl', 'people'];
  const list = data ? Object.keys(data).filter(item => !excludedList.includes(item)) : [];

  const props = {
    title: data?.name,
    card: list.reduce((initial, item) => ( { ...initial, [item]: data[item] } ), {} ),
  };

  return { data: props, error, isValidating };
};
