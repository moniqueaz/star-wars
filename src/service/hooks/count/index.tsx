import useSWR from 'swr';
import { fetcher } from 'utils/helpers';
import config from 'infra/config';

export const useCount = (slug: string | string[]) => {
  const { data, isValidating, error } = useSWR(
      `${config.get}/${slug}`,
      fetcher,
  );
  const count = data?.count || 0;
  return { count, isValidating, error };
};

