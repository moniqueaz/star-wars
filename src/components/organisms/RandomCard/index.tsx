import * as S from './styles';
import { useCard } from 'service/hooks/card';
import Card from 'components/molecules/Card';
import Button from 'components/atoms/Button';
import Placeholder from './placeholder';

type RandomCardProps = {
  count: number,
  callback?: () => any,
  slug: string | string[]
}

const RandomCard = ( { count, callback, slug }:RandomCardProps) => {
  const { data, error, isValidating } = useCard(slug, count);

  if (!!error) {
    return <div>Error</div>;
  }

  return (
    <S.Wrapper>
      <S.Content>
        {
        isValidating ? <Placeholder /> : (
          <Card title={data.title} data={data.card} image="https://via.placeholder.com/250"/>
          )
        }
      </S.Content>
      <Button color="secondary" callback={callback}>Random</Button>
    </S.Wrapper>
  );
};

export default RandomCard;
