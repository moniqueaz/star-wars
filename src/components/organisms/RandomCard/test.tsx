import { render, screen } from '@testing-library/react';
import * as Card from 'service/hooks/card';
import RandomCard from '.';

jest.mock('service/hooks/card');

jest.mock('components/atoms/Button', () => {
  return {
    __esModule: true,
    default: function Mock() {
      return <div data-testid="Mock Button"/>;
    },
  };
} );

jest.mock('components/molecules/Card', () => {
  return {
    __esModule: true,
    default: function Mock() {
      return <div data-testid="Mock Card"/>;
    },
  };
} );

describe('<RandomCard />', () => {
  it('should render button and card', () => {
    jest.spyOn(Card, 'useCard').mockReturnValue( {
      data: { title: 'Titulo', card: { nave: 'nave' } }, error: '',
      isValidating: false,
    } );
    const { container } = render(<RandomCard count={5} slug="people"/>);
    expect(screen.getByTestId('Mock Button')).toBeInTheDocument();
    expect(screen.getByTestId('Mock Card')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  } );

  it('should render isValidationg', () => {
    jest.spyOn(Card, 'useCard').mockReturnValue( {
      data: { title: '', card: {} }, error: 'error message',
      isValidating: false,
    } );
    render(<RandomCard count={5} slug="people"/>);
    expect(screen.getByText(/Error/i)).toBeInTheDocument();
  } );

  it('should render error', () => {
    jest.spyOn(Card, 'useCard').mockReturnValue( {
      data: { title: '', card: {} }, error: '',
      isValidating: true,
    } );
    render(<RandomCard count={5} slug="people"/>);
    expect(screen.getByText(/Loading/i)).toBeInTheDocument();
  } );
} );

