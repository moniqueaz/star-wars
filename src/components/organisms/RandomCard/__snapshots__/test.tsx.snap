// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RandomCard /> should render button and card 1`] = `
<div
  class=""
>
  <div
    class=""
  >
    <div
      data-testid="Mock Card"
    />
  </div>
  <div
    data-testid="Mock Button"
  />
</div>
`;
